pkg/orchestrator/controller/controller.go

```diff
@@ -0,0 +1,264 @@
+	if err != nil {
+		klog.Errorf("error building submariner clientset for cluster %s: %v", clusterID, err)
+		return
+	}
+
+	c.Lock()
+	defer c.Unlock()
+
+	c.removeExistingClusterWatcher(clusterID)
+
+	watcher := &clusterWatcher{
+		clusterID: clusterID,
+		stopChan:  make(chan struct{}),
+		clusterWorkQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("Cluster watcher for %s", clusterID)),
+		endpointWorkQueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			fmt.Sprintf("Endpoint watcher for %s", clusterID)),
+	}
+	c.clusterWatchers[clusterID] = watcher
+
+	submarinerInformerFactory := submarinerInformers.NewSharedInformerFactoryWithOptions(submarinerClient, 0,
+		submarinerInformers.WithNamespace(submarinerNamespace))
+
+	submV1 := submarinerInformerFactory.Submariner().V1()
+	clusterInformer := submV1.Clusters().Informer()
+	clusterInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.clusterWorkQueue))
+
+	endpointInformer := submV1.Endpoints().Informer()
+	endpointInformer.AddEventHandler(newResourceEventHandler(clusterID, watcher.endpointWorkQueue))
+
+	submarinerInformerFactory.Start(watcher.stopChan)
+
+	go runQueueWorker(clusterID, c.federator, clusterInformer, watcher.clusterWorkQueue)
+	go runQueueWorker(clusterID, c.federator, endpointInformer, watcher.endpointWorkQueue)
+}
+
+func getLabel(obj interface{}, key string) (string, bool) {
+	if deleted, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+		obj = deleted.Obj
+	}
+
+	metadata, err := meta.Accessor(obj)
+	if err != nil {
+		return "", false
+	}
+
+	labels := metadata.GetLabels()
+	if labels == nil {
+		return "", false
+	}
+
+	label, exists := labels[key]
+	return label, exists
+}
+
+func setLabel(obj runtime.Object, key, value string) runtime.Object {
+	copy := obj.DeepCopyObject()
+
+	// We can safely ignore the error as all the resources implement meta.Object
+	metadata, _ := meta.Accessor(copy)
+
+	labels := metadata.GetLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	labels[key] = value
+	metadata.SetLabels(labels)
+	return copy
+}
+
+func newResourceEventHandler(clusterID string, workQueue workqueue.RateLimitingInterface) cache.ResourceEventHandler {
+	enqueue := func(obj interface{}, keyFunc func(obj interface{}) (string, error)) {
+		key, err := keyFunc(obj)
+		if err != nil {
+			klog.Errorf("Error creating MetaNamespaceKey: %v", err)
+			return
+		}
+
+		label, exists := getLabel(obj, clusterIDLabelKey)
+		if exists && label != clusterID {
+			klog.V(2).Infof("Label \"%s=%s\" for \"%s\" does not match \"%s\" - skipping", clusterIDLabelKey,
+				label, key, clusterID)
+			return
+		}
+
+		workQueue.Add(key)
+	}
+
+	return &cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			enqueue(obj, cache.MetaNamespaceKeyFunc)
+		},
+		UpdateFunc: func(old interface{}, new interface{}) {
```

Comment:
Since you're using ResourceEventHandlerFuncs you don't have to specify the UpdateFunc since it does nothign by default (the TODO can stay above the return).